src/lib.rs

```diff
@@ -137,117 +129,245 @@ impl Context {
+            return Err(failure::err_msg(
+                "Only one of `--bin`, `--example` or `--lib` must be specified",
+            ));
+        }
+
+        if bin {
+            (
+                Some(Artifact::Bin(matches.value_of("bin").unwrap())),
+                release,
+            )
+        } else if example {
+            (
+                Some(Artifact::Example(matches.value_of("example").unwrap())),
+                release,
+            )
+        } else if lib {
+            (Some(Artifact::Lib), release)
+        } else {
+            (None, release)
+        }
+    } else {
+        (None, false)
+    };
+
+    let mut cargo = Command::new("cargo");
+    cargo.arg("build");
+
+    // NOTE we do *not* use the `build_target` info here because Cargo will figure things out on
+    // its own (i.e. it will search and parse .cargo/config, etc.)
+    if let Some(target) = target_flag {
+        cargo.args(&["--target", target]);
     }
 
-    /* Private API */
-    fn bindir(&self) -> &Path {
-        &self.bindir
+    match artifact {
+        Some(Artifact::Bin(bin)) => {
+            cargo.args(&["--bin", bin]);
+        }
+        Some(Artifact::Example(example)) => {
+            cargo.args(&["--example", example]);
+        }
+        Some(Artifact::Lib) => {
+            cargo.arg("--lib");
+        }
+        None => {}
     }
 
-    fn demangle<'i>(&self, input: &'i str) -> Cow<'i, str> {
-        self.re.replace_all(input, |cs: &Captures| {
-            format!("{}", rustc_demangle::demangle(cs.get(0).unwrap().as_str()))
-        })
+    if release {
+        cargo.arg("--release");
     }
 
-    #[cfg(unused)]
-    fn host(&self) -> &str {
-        &self.host
+    if artifact.is_some() {
+        if verbose {
+            eprintln!("{:?}", cargo);
+        }
+
+        let status = cargo.status()?;
+
+        if !status.success() {
+            return Ok(status.code().unwrap_or(1));
+        }
     }
 
-    fn target(&self) -> &str {
-        self.target.as_ref().unwrap_or(&self.host)
+    let mut tool_args = vec![];
+    if let Some(arg) = matches.value_of("--") {
+        tool_args.push(arg);
     }
 
-    fn tool(&self, name: &str) -> Command {
-        Command::new(self.bindir().join(&*exe(name)))
+    if let Some(args) = matches.values_of("args") {
+        tool_args.extend(args);
     }
-}
 
-/// Shared entry point for the Cargo subcommands
-pub fn run<F, G>(demangle: bool, tool: F, post_process: G) -> Result<i32>
-where
-    F: FnOnce(&Context) -> Command,
-    G: for<'s> FnOnce(&Context, &'s str) -> Cow<'s, str>,
-{
-    let ctxt = Context::new()?;
-    let mut tool = tool(&ctxt);
+    let ctxt = Context::new(target_flag)?;
 
-    tool.args(ctxt.tool_args());
+    let mut lltool = ctxt.tool(tool, ctxt.target());
 
-    let stderr = io::stderr();
-    let mut stderr = stderr.lock();
+    if let Some(kind) = artifact {
+        let artifact = cargo::artifact(kind, release, target_flag, ctxt.build_target())?;
```

Comment:
Would need to be: `let artifact = cargo::artifact(&kind, release, target_flag, ctxt.build_target())?;` to match up with clippy suggestion